profiling/src/profile/mod.rs

```diff
@@ -430,14 +434,68 @@ impl Profile {
         // size of 32KiB should definitely out-perform starting at zero for
         // time consumed, allocator pressure, and allocator fragmentation.
         const INITIAL_PPROF_BUFFER_SIZE: usize = 32 * 1024;
-        let mut buffer: Vec<u8> = Vec::with_capacity(INITIAL_PPROF_BUFFER_SIZE);
-        profile.encode(&mut buffer)?;
+        let mut encoder = CompressedProtobufSerializer::with_capacity(INITIAL_PPROF_BUFFER_SIZE);
+
+        for (sample, timestamp, mut values) in std::mem::take(&mut self.observations).into_iter() {
+            let labels = self.translate_and_enrich_sample_labels(sample, timestamp)?;
+            let location_ids: Vec<_> = self
+                .get_stacktrace(sample.stacktrace)
+                .locations
+                .iter()
+                .map(Id::to_raw_id)
+                .collect();
+            self.upscaling_rules
+                .upscale_values(&mut values, &labels, &self.sample_types)?;
+
+            let item = pprof::Sample {
+                location_ids,
+                values,
+                labels,
+            };
+
+            encoder.encode(ProfileSamplesEntry::from(item))?;
+        }
+
+        // We need to do this out of order because we use the sample_types while
+        // upscaling
+        for sample_type in self.sample_types.into_iter() {
+            let item: pprof::ValueType = sample_type.into();
+            encoder.encode(ProfileSampleTypesEntry::from(item))?;
+        }
```

Comment:
Sorry, what is out of order here? Aren't the sample types the first field in a profile?

Reply:
We first emitted the `Sample`, then the `SampleTypes`, which is the opposite order of the profile.  I'll try to clean up the comment to clarify.